src/v/panda_link/service.cc

```diff
@@ -0,0 +1,170 @@
+
+ss::future<> service::start() {
+    vlog(pllog.info, "Starting panda link service");
+    _manager = std::make_unique<manager>(
+      _self,
+      std::make_unique<link_registry_adapter>(&_plf->local()),
+      std::make_unique<default_link_factory>());
+
+    co_await _manager->start();
+    register_notifications();
+}
+
+ss::future<> service::stop() {
+    vlog(pllog.info, "Stopping panda link service");
+    unregister_notifications();
+    co_await _manager->stop();
+}
+
+void service::register_notifications() {
+    auto pl_notif_id = _plf->local().register_for_updates(
+      [this](model::id_t id) { _manager->on_link_change(id); });
+    _notification_cleanups.emplace_back([this, pl_notif_id] {
+        _plf->local().unregister_for_updates(pl_notif_id);
+    });
+
+    auto leadership_notif_id
+      = _group_manager->local().register_leadership_notification(
+        [this](
+          raft::group_id group_id,
+          ::model::term_id term,
+          std::optional<::model::node_id> leader) {
+            on_leadership_change(group_id, term, leader);
+        });
+    _notification_cleanups.emplace_back([this, leadership_notif_id] {
+        _group_manager->local().unregister_leadership_notification(
+          leadership_notif_id);
+    });
+
+    auto umanage_notif_id
+      = _partition_manager->local().register_unmanage_notification(
+        ::model::kafka_namespace, [this](::model::topic_partition_view tp) {
+            on_unmanage_notification(tp);
+        });
+    _notification_cleanups.emplace_back([this, umanage_notif_id] {
+        _partition_manager->local().unregister_unmanage_notification(
+          umanage_notif_id);
+    });
+    auto manage_notif_id
+      = _partition_manager->local().register_manage_notification(
+        ::model::kafka_namespace,
+        [this](const ss::lw_shared_ptr<cluster::partition>& p) {
+            on_manage_notification(p);
+        });
+    _notification_cleanups.emplace_back([this, manage_notif_id] {
+        _partition_manager->local().unregister_manage_notification(
+          manage_notif_id);
+    });
+}
+
+void service::unregister_notifications() { _notification_cleanups.clear(); }
+
+void service::on_leadership_change(
+  raft::group_id group_id,
+  ::model::term_id term,
+  std::optional<::model::node_id> leader) {
+    vlog(
+      pllog.trace,
+      "on_leadership_change: group_id={}, term={}, leader={}",
+      group_id,
+      term,
+      leader);
+    auto partition = _partition_manager->local().partition_for(group_id);
+    if (!partition) {
+        vlog(
+          pllog.debug,
+          "got leadership notification for unknown partition: {}",
+          group_id);
+        return;
+    }
+    bool node_is_leader = leader.has_value() && *leader == _self;
+    if (!node_is_leader) {
+        _manager->on_leadership_change(partition->ntp(), ntp_leader::no);
+    }
+    auto is_leader = partition->is_elected_leader() ? ntp_leader::yes
```

Comment:
this should be `is_leader()` `is_elected_leader` returns true right after election but before the term is confirmed